### ISSUE TITLE
Vim9: should we enforce <stack> instead of <sfile> inside a function?

### ISSUE THREAD

**Issue by lacygoill** [vim/vim#9189]
`<sfile>` is too ambiguous in legacy.  At the script level, it gives the path of the sourced script, but in a function it gives the call stack.  We can't remove this ambiguity in legacy because it could break existing scripts, but we could remove it in Vim9, by enforcing `<stack>` instead of `<sfile>` in a function.  Vim supports the latter since [8.2.1297](https://github.com/vim/vim/releases/tag/v8.2.1297).

This would make the code more readable, because `<sfile>` reads as if we get a simple file path, while in reality we get a call stack.  Besides, it would make issues such as https://github.com/vim/vim/issues/9178 easier to debug.  The solution is obvious in hindsight, but the debugging took more time than it should have, because the fact that `<sfile>` gives a call stack might be hidden when we use filename modifiers.

So this snippet:
```vim
vim9script
def Func()
    eval expand('<sfile>')
enddef
Func()
echo 'no error'
```
    no error

Could give an error such as:

    E1234: <sfile> can only be used at script-level

**Comment by brammool** [comment]
What if someone wants to get the script name where the function was defined?
I can't find something simpler than setting a variable at script level that expands <sfile>
and using that variable in the function.
Perhaps we can add "<funcfile>", as the file where the function was defined.

**Comment by lacygoill** [comment]
> What if someone wants to get the script name where the function was defined?
   > I can't find something simpler than setting a variable at script level that expands
   > and using that variable in the function.

Yes, that's exactly what I do.  I don't know any other way.

   > Perhaps we can add "", as the file where the function was defined.

Sorry, the name between angle brackets has disappeared on github.  But yes, I guess we could introduce another angle-bracket expression such as `<script>` which would give us the name of the script from everywhere (script level and function).

**Comment by brammool** [comment]
I was thinking of "funcfile", but "script" would be simpler.  So you would stop using "sfile", which has a value that depends on the place where it is used, and use "script" and "stack" instead.

**Comment by lacygoill** [comment]
Closing because the original issue has been fixed in patch 8.2.3646.  The `<script>` angle-bracket expression is a separate issue which has been noted in the [todo list](https://github.com/vim/vim/blob/59f4f9505ae7ca2499904b94100db103e5ada5a6/runtime/doc/todo.txt#L232).
